gc/include/omtalk/MutatorMutex.h

```diff
@@ -0,0 +1,117 @@
+#ifndef OMTALK_MUTATOR_MUTEX_H
+#define OMTALK_MUTATOR_MUTEX_H
+
+#include <condition_variable>
+#include <cstdint>
+#include <mutex>
+#include <omtalk/Util/Annotations.h>
+
+namespace omtalk::gc {
+
+/// Private mutex data
+struct MutatorMutexData {
+  /// 0 indicates shared state
+  /// 1 indicates exclusive state
+  std::uint8_t state;
+  std::uintptr_t count;
+};
+
+/// Provides functionality similiar to a reader/writer lock.  Mutators all have
+/// shared access to the mutex.  When exclusive access is requested all mutator
+/// threads should advance to the next GC safepoint and yield.
+class OMTALK_MUTEX_CAPABILITY MutatorMutex {
+public:
+  MutatorMutex() noexcept {}
+
+  ~MutatorMutex() noexcept {}
+
+  /// Attach a running mutator
+  void attach() noexcept OMTALK_ACQUIRE_SHARED() {
+    std::unique_lock lock(mutex);
+    auto check = [this] { return data.state == 0; };
+    yieldCV.wait(lock, check);
+    data.count++;
+  }
+
+  /// Detach a running mutator
+  void detach() noexcept OMTALK_RELEASE_SHARED() {
+    std::unique_lock lock(mutex);
+    data.count--;
+    if (data.count == 0) {
+      requestCV.notify_all();
+    }
+  }
+
+  /// Yield shared access and block until shared access is regained.
+  bool yield() noexcept {
+    if (!requested()) {
+      return false;
+    }
+    std::unique_lock lock(mutex);
+    data.count--;
+    if (data.count == 0) {
+      requestCV.notify_all();
+    }
+    auto check = [this] { return data.state == 0; };
+    yieldCV.wait(lock, check);
+    data.count++;
+    return true;
+  }
+
+  /// Returns true if access was requested.
+  bool requested() noexcept {
+    std::scoped_lock lock(mutex);
+    return data.state;
```

Comment:
TSAN reports a data race when I switch this to an atomic read.

Reply:
because you are not using an atomic write